<center><h1><strong>WeRateDogs Data Wrangling</strong></h1></center>

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#WeRateDogs-Data-Wrangling" data-toc-modified-id="WeRateDogs-Data-Wrangling-1"><strong>WeRateDogs Data Wrangling</strong></a></span></li><li><span><a href="#Introduction" data-toc-modified-id="Introduction-2">Introduction</a></span></li><li><span><a href="#1.-Gathering" data-toc-modified-id="1.-Gathering-3">1. Gathering</a></span><ul class="toc-item"><li><span><a href="#1.1-The-WeRateDogs-Twitter-Archive" data-toc-modified-id="1.1-The-WeRateDogs-Twitter-Archive-3.1">1.1 The WeRateDogs Twitter Archive</a></span></li><li><span><a href="#1.2-The-Tweet-image-predictions" data-toc-modified-id="1.2-The-Tweet-image-predictions-3.2">1.2 The Tweet image predictions</a></span></li><li><span><a href="#1.3-Twitter-API" data-toc-modified-id="1.3-Twitter-API-3.3">1.3 Twitter API</a></span></li><li><span><a href="#1.4-Gathering-Output" data-toc-modified-id="1.4-Gathering-Output-3.4">1.4 Gathering Output</a></span></li></ul></li><li><span><a href="#2.-Assessing-the-Data" data-toc-modified-id="2.-Assessing-the-Data-4">2. Assessing the Data</a></span><ul class="toc-item"><li><span><a href="#2.1-Twitter-Archive" data-toc-modified-id="2.1-Twitter-Archive-4.1">2.1 Twitter Archive</a></span></li><li><span><a href="#2.2-Tweet-Image-Predictions" data-toc-modified-id="2.2-Tweet-Image-Predictions-4.2">2.2 Tweet Image Predictions</a></span></li><li><span><a href="#2.3-Twitter-API" data-toc-modified-id="2.3-Twitter-API-4.3">2.3 Twitter API</a></span></li><li><span><a href="#2.4-Assessing-Output" data-toc-modified-id="2.4-Assessing-Output-4.4">2.4 Assessing Output</a></span></li></ul></li><li><span><a href="#3.-Cleaning" data-toc-modified-id="3.-Cleaning-5">3. Cleaning</a></span><ul class="toc-item"><li><span><a href="#3.1-Twitter-Archive" data-toc-modified-id="3.1-Twitter-Archive-5.1">3.1 Twitter Archive</a></span></li><li><span><a href="#3.2-Tweet-Image-Predictions" data-toc-modified-id="3.2-Tweet-Image-Predictions-5.2">3.2 Tweet Image Predictions</a></span></li><li><span><a href="#3.3-Twitter-API" data-toc-modified-id="3.3-Twitter-API-5.3">3.3 Twitter API</a></span></li><li><span><a href="#3.4-Merging-Data" data-toc-modified-id="3.4-Merging-Data-5.4">3.4 Merging Data</a></span></li><li><span><a href="#3.5-Storing-Data" data-toc-modified-id="3.5-Storing-Data-5.5">3.5 Storing Data</a></span></li></ul></li></ul></div>

# Introduction

This project started as an assessment for the **Udacity Data Analysis Professional Track**.

The dataset is the tweet archive of Twitter user [@dog_rates](https://twitter.com/dog_rates), also known as 
[WeRateDogs](https://en.wikipedia.org/wiki/WeRateDogs). WeRateDogs is a Twitter account that rates people's dogs with a humorous comment about the dog. These ratings almost always have a denominator of 10. The numerators, though? Almost always greater than 10. 11/10, 12/10, 13/10, etc. Why? Because ["they're good dogs Brent."](https://knowyourmeme.com/memes/theyre-good-dogs-brent) WeRateDogs has over 4 million followers and has received international media coverage.

<center><img width="50%" src ="dog-rates-social.jpg"/></center>

The data wrangling process was achieved using python and the following python packages and libraries:

- [NumPy](https://numpy.org/)
- [Panadas](https://pandas.pydata.org/)
- [Requests](https://requests.readthedocs.io/en/master/)
- [Tweepy](https://www.tweepy.org/)
- [json](https://docs.python.org/3/library/json.html)
- [re](https://docs.python.org/3/library/re.html)

# 1. Gathering

Three sources of data were gathered for this Project:
1. Twitter archive provided by WeRateDogs
2. Image predictions from the neural network data downloaded programmatically.
3. Information from the Twitter API scraped programmatically.

We begin by importing the required libraries and packages that will help us in our data wrangling and analysis

In [1]:
#import required libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import os
import requests
import tweepy
import json
import re

## 1.1 The WeRateDogs Twitter Archive

- The 'twitter-archive-enhanced.csv' file was provided by Udacity and downloaded manually.
- Then it was easily read into a pandas DataFrame named `df_archive`.

In [2]:
df_archive = pd.read_csv('twitter-archive-enhanced.csv')
df_archive.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None  
2                  10   Archie  None    None   None  None

## 1.2 The Tweet image predictions

- The 'image-predictions.tsv' shows the results of a neural network that detects what breed of dog (or other object, animal, etc.) is present in each tweet.   
- The file was downloaded programmatically using the Requests python library and then saved to a local file.
- The file name was extracted from the url using the split function.
- The file was then read into a pandas DataFrame `df_pred` while specifying that it's a tsv by setting the `sep` parameter to `'\t'`.

In [3]:
# Download image predictions file if it doesn't exist
pred_url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
file_name = pred_url.split('/')[-1]

# only download the file if the file doesn't already exist on disk.
if os.path.exists(file_name):
    print(f"File {file_name} already exists")
else:
    r = requests.get(pred_url)
    # write the reponse to a local file
    with open('image-predictions.tsv','wb') as file:
        file.write(r.content)
        print(f"successfully downloaded {file_name} and written to disc")

File image-predictions.tsv already exists


In [4]:
# Read the image predictions file into a DataFrame
df_pred = pd.read_csv('image-predictions.tsv',sep='\t')
df_pred.sample(3)

tweet_id                                          jpg_url  \
663  682662431982772225  https://pbs.twimg.com/media/CXlN1-EWMAQdwXK.jpg   
894  699434518667751424  https://pbs.twimg.com/media/CbTj--1XEAIZjc_.jpg   
710  685307451701334016  https://pbs.twimg.com/media/CYKzfTTWMAEeTN7.jpg   

     img_num                p1   p1_conf  p1_dog                 p2   p2_conf  \
663        1            beagle  0.413824    True           Cardigan  0.263553   
894        1  golden_retriever  0.836572    True             kuvasz  0.105946   
710        1        Pomeranian  0.963176    True  Shetland_sheepdog  0.019468   

     p2_dog                  p3   p3_conf  p3_dog  
663    True              basset  0.167618    True  
894    True  Labrador_retriever  0.025144    True  
710    True            keeshond  0.008605    True

## 1.3 Twitter API 

- A Twitter Developer account was created and used to query more data using the tweet ids from the `df_archive` DataFrame.
- The Twitter API was used to query each tweet's JSON data using Python's Tweepy library and stored each tweet's entire set of JSON data as a line in a file called tweet_json.txt file.
- The 'tweet._json.txt' file line by line into a pandas DataFrame `df_api` with the following columns:
    - tweet ID.
    - retweet count. 
    - favorite count. 

- Note: Twitter API keys, secrets, and tokens were read from an external file and were not submitted with the project.

In [5]:
tweet_ids = df_archive['tweet_id']
json_file = 'tweet_json.txt'


# only Query the API if the tweet_json.txt file doesn't exist
if os.path.exists(json_file):
    print(f"File {json_file} already exists")
else:
    # import the keys files where the API keys are stored and it is not submitted with the project
    import keys
    # Twitter API authorization 
    auth = tweepy.OAuthHandler(keys.consumer_key, keys.consumer_secret)
    auth.set_access_token(keys.access_token, keys.access_secret)

    api = tweepy.API(auth)
    
    count = 0
    fails_dict = {}
    # Save each tweet's returned JSON as a new line in a .txt file
    with open(json_file, 'w') as outfile:
        # Loop through each tweet and write its information into a separate line of the .txt file
        for tweet_id in tweet_ids:
            count += 1
            print(str(count) + ": " + str(tweet_id))
            try:
                tweet = api.get_status(tweet_id, tweet_mode='extended')
                print("Success")
                json.dump(tweet._json, outfile)
                outfile.write('\n')
            except tweepy.TweepError as e:
                print("Fail")
                fails_dict[tweet_id] = e
                pass
    print(f"File {json_file} successfully written to Disk")
    print("Some tweets returned the following errors")
    print(fails_dict)



File tweet_json.txt already exists


In [6]:
# read .txt file line by line into a pandas DataFrame

labels = ['tweet_id','retweet_count','favorite_count']

df_api = pd.DataFrame(columns=labels)

with open(json_file) as file:
    for line in file:
        data = json.loads(line)
        df_api = df_api.append({
            'tweet_id':data['id'],
            'retweet_count':data['retweet_count'],
            'favorite_count':data['favorite_count']
            },ignore_index=True)
        
df_api.sample(5)

tweet_id retweet_count favorite_count
465   815639385530101762          1633           8164
1314  670040295598354432           104            706
1371  668872652652679168           279            493
717   781163403222056960          2720           9727
691   785170936622350336          4774          11947

## 1.4 Gathering Output
 
The three files were successfully downloaded, written to a local file and read into the following pandas DataFrames:
- `df_archive`:  Data from the WeRateDogs Twitter archive.
- `df_pred`:  Data from the image predictions from the neural network.
- `df_api`:  Data scraped from the Twitter API.


# 2. Assessing the Data

We start assessing the data to try to find any Quality or tidiness issues. 

The aim is to asses all three DataFrames with respect to Quality dimensions:
 - Completeness
 - Validity
 - Accuracy
 - Consistency
 
and the following tidiness rules:
- Each variable forms a column.
- Each observation forms a row.
- Each type of observational unit forms a table.

## 2.1 Twitter Archive

Using the `info` method we try to find missing data and look to see if we have non-original tweets (retweets and replies). Also check the is we already have the correct data types for all columns.

In [7]:
df_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [8]:
df_archive.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2327  666407126856765440                    NaN                  NaN   
940   753026973505581056                    NaN                  NaN   
1377  701601587219795968                    NaN                  NaN   
2131  670086499208155136                    NaN                  NaN   
475   816062466425819140                    NaN                  NaN   

                      timestamp  \
2327  2015-11-17 00:06:54 +0000   
940   2016-07-13 00:43:15 +0000   
1377  2016-02-22 02:57:08 +0000   
2131  2015-11-27 03:47:25 +0000   
475   2017-01-02 23:23:48 +0000   

                                                 source  \
2327  <a href="http://twitter.com/download/iphone" r...   
940   <a href="http://twitter.com/download/iphone" r...   
1377  <a href="http://twitter.com/download/iphone" r...   
2131  <a href="http://twitter.com/download/iphone" r...   
475   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2327  This is a southern Vesuvius bumblegruff. Can d...                  NaN   
940   Say hello to Tayzie. She's a Barbadian Bugaboo...                  NaN   
1377  This is Coopson. He's a Blingin Schnitzel. Bui...                  NaN   
2131  "Hi yes this is dog. I can't help with that s-...                  NaN   
475   RT @dog_rates: Meet Jack. He's one of the rare...         8.159907e+17   

      retweeted_status_user_id retweeted_status_timestamp  \
2327                       NaN                        NaN   
940                        NaN                        NaN   
1377                       NaN                        NaN   
2131                       NaN                        NaN   
475               4.196984e+09  2017-01-02 18:38:42 +0000   

                                          expanded_urls  rating_numerator  \
2327  https://twitter.com/dog_rates/status/666407126...                 7   
940   https://twitter.com/dog_rates/status/753026973...                10   
1377  https://twitter.com/dog_rates/status/701601587...                10   
2131  https://twitter.com/dog_rates/status/670086499...                10   
475   https://www.gofundme.com/surgeryforjacktheminp...                11   

      rating_denominator     name doggo floofer pupper puppo  
2327                  10        a  None    None   None  None  
940                   10   Tayzie  None    None   None  None  
1377                  10  Coopson  None    None   None  None  
2131                  10     None  None    None   None  None  
475                   10     Jack  None    None   None  None

In [9]:
# check for duplicated entries
sum(df_archive['tweet_id'].duplicated())

0

It si clear that we have 181 tweets that are retweets and 78 tweets that are replies. That means these are not original tweets and will need to be filtered out.

Next, we start exploring the ratings data.

In [10]:
df_archive['rating_denominator'].value_counts().keys()

Int64Index([10, 11, 50, 80, 20, 2, 16, 40, 70, 15, 90, 110, 120, 130, 150, 170,
            7, 0],
           dtype='int64')

In [11]:
df_archive[df_archive['rating_denominator'] != 10].sample(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1663  682808988178739200           6.827884e+17         4.196984e+09   
1254  710658690886586372                    NaN                  NaN   
1598  686035780142297088           6.860340e+17         4.196984e+09   

                      timestamp  \
1663  2016-01-01 06:22:03 +0000   
1254  2016-03-18 02:46:49 +0000   
1598  2016-01-10 04:04:10 +0000   

                                                 source  \
1663  <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
1598  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1663  I'm aware that I could've said 20/16, but here...                  NaN   
1254  Here's a brigade of puppers. All look very pre...                  NaN   
1598  Yes I do realize a rating of 4/20 would've bee...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1663                       NaN                        NaN   
1254                       NaN                        NaN   
1598                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1663                                                NaN                20   
1254  https://twitter.com/dog_rates/status/710658690...                80   
1598                                                NaN                 4   

      rating_denominator  name doggo floofer pupper puppo  
1663                  16  None  None    None   None  None  
1254                  80  None  None    None   None  None  
1598                  20  None  None    None   None  None

In [12]:
df_archive['rating_numerator'].value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

The rating denominator was not always equal to 10 and also the rating numerator had some inconsistent values either too small or too large. After examining the data set by filtering the illogical and exploring both their text and their images, it was deduced that the reasons for the issues in these ratings were:
- The ratings of pack of dogs is added together making the denominator multiples of 10.
- The extraction process assumed that '/' character only appears once in the text and that it always indicates rating which was found not to be true.


Next we begin exploring the 'name' column as it was noticed from the visual assessment that it contained some words that are not considered to be actual names. 

In [13]:
df_archive['name'].value_counts()

None           745
a               55
Charlie         12
Cooper          11
Oliver          11
              ... 
Gunner           1
Sully            1
Rodney           1
Brownie          1
Shakespeare      1
Name: name, Length: 957, dtype: int64

In [14]:
names = df_archive['name'].value_counts().keys()
not_names = []

for name in names:
    if name.islower() or len(name) < 2:
        not_names.append(name)
print(not_names)

['a', 'the', 'an', 'very', 'one', 'just', 'quite', 'actually', 'getting', 'mad', 'not', 'unacceptable', 'incredibly', 'all', 'this', 'my', 'infuriating', 'light', 'life', 'old', 'such', 'space', 'by', 'his', 'O', 'officially']


The name column has a lot of missing values that are indicated as "None" in a string format.
Also, it contained a lot of words that should not be considered names. that probably happened because the extraction process assumed that names only and must come after the words "this is" which is not always true. 

According to the *Dogtionary*, doggo, floofer, pupper and puppo are all dog stages. We have to explore these columns together to get better insight about the data.

In [15]:
df_archive[['doggo','floofer','pupper','puppo']].value_counts()

doggo  floofer  pupper  puppo
None   None     None    None     1976
                pupper  None      245
doggo  None     None    None       83
None   None     None    puppo      29
doggo  None     pupper  None       12
None   floofer  None    None        9
doggo  floofer  None    None        1
       None     None    puppo       1
dtype: int64

- The dog stages columns ('doggo', 'flooofer', 'pupper', 'puppo') are values and not variables. They all represent different dog stages. They should be one column instead.
- Some entries has more than one dog stage:
    - 12 entry has both doggo and pupper.
    - 1 entry has both doggo and floofer.
- Missing values are written as "None" in a string format instead on nulls.

In [16]:
df_archive['source'].value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

The 'source' column contains the whole HTML 'a' tag while the source is only the inner text of the tag.

## 2.2 Tweet Image Predictions

In [17]:
df_pred.sample(5)

tweet_id                                          jpg_url  \
874   698178924120031232  https://pbs.twimg.com/media/CbBuBhbWwAEGH29.jpg   
1665  812503143955202048  https://pbs.twimg.com/media/C0aXTLqXEAADxBi.jpg   
329   672082170312290304  https://pbs.twimg.com/media/CVO3KodXAAAj1de.jpg   
1653  809920764300447744  https://pbs.twimg.com/media/Cz1qo05XUAQ4qXp.jpg   
489   675522403582218240  https://pbs.twimg.com/media/CV_wCh8W4AEWWZ9.jpg   

      img_num                        p1   p1_conf  p1_dog                p2  \
874         1  Chesapeake_Bay_retriever  0.351868    True          malinois   
1665        2                     loupe  0.546856   False          web_site   
329         1                   hamster  0.132440   False        toy_poodle   
1653        1           Norwich_terrier  0.397163    True        toy_poodle   
489         1            cocker_spaniel  0.299708    True  golden_retriever   

       p2_conf  p2_dog                  p3   p3_conf  p3_dog  
874   0.207753    True  Labrador_retriever  0.154606    True  
1665  0.345298   False              bubble  0.010528   False  
329   0.123962    True              bubble  0.056212   False  
1653  0.274540    True    miniature_poodle  0.134667    True  
489   0.263665    True        Irish_setter  0.080323    True

In [18]:
sum(df_pred['tweet_id'].duplicated())

0

In [19]:
df_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [20]:
df_pred['img_num'].value_counts()

1    1780
2     198
3      66
4      31
Name: img_num, dtype: int64

- prediction levels are separated in different columns while they actually all represent one variable "prediction_level" 
- Even though some tweets have more than one image but no duplicates were found. That means that even though tweets might contain multiple images, we only have one image per tweet for prediction.

## 2.3 Twitter API

In [21]:
df_api

tweet_id retweet_count favorite_count
0     892420643555336193          7461          35352
1     892177421306343426          5537          30593
2     891815181378084864          3664          23025
3     891689557279858688          7637          38647
4     891327558926688256          8228          36916
...                  ...           ...            ...
1527  666049248165822465            39             95
1528  666044226329800704           124            264
1529  666033412701032449            39            108
1530  666029285002620928            41            118
1531  666020888022790149           448           2356

[1532 rows x 3 columns]

In [22]:
df_api.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1532 entries, 0 to 1531
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   tweet_id        1532 non-null   object
 1   retweet_count   1532 non-null   object
 2   favorite_count  1532 non-null   object
dtypes: object(3)
memory usage: 36.0+ KB


- The 'retweet_count' and 'favorite_count' columns have the wrong datatype. 
- The index column here is the same as the `df_archive` DataFrame and they should be merged together.

## 2.4 Assessing Output

After the data was assessed both visually and programmatically, The following issues were concluded:

### Quality

#### Completeness: 
do we have all of the records that we should? Do we have missing records or not? Are there specific rows, columns, or cells missing?

`df_archive`
- 'expanded_urls' have missing values. that means they don't have image information.

`df_api`
- The API query failed to retrieve data for 25 tweets because they are either no longer available or access is denied.


#### Validity: 
we have the records, but they're not valid, i.e., they don't conform to a defined schema. A schema is a defined set of rules for data. These rules can be real-world constraints (e.g. negative height is impossible) and table-specific constraints (e.g. unique key constraints in tables).

`df_archive`

- The 'is_reply_status_id' and 'retweeted_status_id' columns have non null values showing that 181 entries are retweets and 78 entries are replies. i.e; not original retweets.
- Also 'expanded_urls' have missing values which means they don't contain images.

#### Accuracy: 
inaccurate data is wrong data that is valid. It adheres to the defined schema, but it is still incorrect. Example: a patient's weight that is 5 lbs too heavy because the scale was faulty.

`df_archive`

- the 'name' column has some values that are not actual name. such as:  
'a', 'the', 'an', 'very', 'one', 'quite', 'just', 'mad', 'actually', 'getting', 'not', 'infuriating', 'space', 'unacceptable', 'his', 'old', 'O', 'life', 'this', 'incredibly', 'such', 'light', 'officially', 'all', 'by', 'my'.
- 12 entries have both lables 'doggo' and 'pupper'
- 1 entry have both labels 'doggo' and 'floofer

#### Consistency: 
inconsistent data is both valid and accurate, but there are multiple correct ways of referring to the same thing. Consistency, i.e., a standard format, in columns that represent the same data across tables and/or within tables is desired.

`df_archive`
- not all 'rating_denominator' values are 10.
- The 'rating_numerator' column has some very low or very large values.
- 'timestamp' and 'retweeted_status_timestamp' column is has type of string. should be datetime.
- the 'source' column values are the complete 'a' tag. Should be reduced to only be actual source values:  
    iPhone, Vine, Web Client and TweetDeck
- The 'name' columns missing values are labeled 'None'.

`df_api`
- the 'tweet_id', 'retweet_count' and 'favorite_count' are of type string. Should be integers.


### Tidiness

`df_archive`
- The dog stages columns ('doggo', 'flooofer', 'pupper', 'puppo') should be one column instead. Because every column should be a variable not values of the same variable.

`df_pred`
- Multiple predictions exist in the same row. the don't follow the tidiness rule 'every row is an observation'. Each prediction should be its own row instead.

`df_api`
- The observation row here is the tweet defined by the 'tweet_id' that is the same as in `df_archive`. The two tables should be one table instead. 

# 3. Cleaning

## 3.1 Twitter Archive

We first copy the original DataFrame to a new one to be cleaned.

In [23]:
df_archive_clean = df_archive.copy()

#### Define
* Convert the 'timestamp' and 'retweeted_status_timestamp' columns from string to datetime using the `pd.to_datetime` function.

#### Code

In [24]:
df_archive_clean['timestamp'] = pd.to_datetime(df_archive_clean['timestamp'])
df_archive_clean['retweeted_status_timestamp'] = pd.to_datetime(df_archive_clean['retweeted_status_timestamp'])

#### Test

In [25]:
# check that the 'timestamp' and 'retweeted_status_timestamp' columns are changed to datetime
df_archive_clean[['timestamp','retweeted_status_timestamp']].dtypes

timestamp                     datetime64[ns, UTC]
retweeted_status_timestamp    datetime64[ns, UTC]
dtype: object

#### Define

* Remove all tweets that are not orignal tweets, i.e; tweets that are retweets or replies. So, the `df_archive_clean` is filtered to only the rows where the columns 'retweeted_status_id' and 'in_reply_to_status_id' are null values.

#### Code

In [26]:
df_archive_clean = df_archive_clean[df_archive_clean['retweeted_status_id'].isnull()]
df_archive_clean = df_archive_clean[df_archive_clean['in_reply_to_status_id'].isnull()]

#### Test

In [27]:
df_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    2097 non-null   int64              
 1   in_reply_to_status_id       0 non-null      float64            
 2   in_reply_to_user_id         0 non-null      float64            
 3   timestamp                   2097 non-null   datetime64[ns, UTC]
 4   source                      2097 non-null   object             
 5   text                        2097 non-null   object             
 6   retweeted_status_id         0 non-null      float64            
 7   retweeted_status_user_id    0 non-null      float64            
 8   retweeted_status_timestamp  0 non-null      datetime64[ns, UTC]
 9   expanded_urls               2094 non-null   object             
 10  rating_numerator            2097 non-null   int64           

#### Define

* Remove the rows with empty 'expanded_urls'. These are the tweets that don't have image information.

#### Code

In [28]:
df_archive_clean = df_archive_clean[df_archive_clean['expanded_urls'].notna()]

#### Test

In [29]:
sum(df_archive_clean['expanded_urls'].isnull())

0

#### Define

- Remove the columns 'in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp' as they are no longer relevant

#### Code

In [30]:
cols = ['in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp',]

df_archive_clean.drop(cols, axis=1, inplace=True)

#### Test

In [31]:
df_archive_clean.columns

Index(['tweet_id', 'timestamp', 'source', 'text', 'expanded_urls',
       'rating_numerator', 'rating_denominator', 'name', 'doggo', 'floofer',
       'pupper', 'puppo'],
      dtype='object')

#### Define

* Combine the dog stages columns ('doggo', 'flooofer', 'pupper', 'puppo') should be one column 'dog_stage' instead. This was done by appliyng a custom funcion `dog_stage` to the whole DataFrame that returns the interpreted dog_stage and convert the 'None' string to a null value
* for cases where 'doggo' and other value exist. 'doggo' will be chosen as 'doggo' is the older dog and dogs can only grow up. So, 'doggo' will definetly be the more valid value.

#### Code

In [32]:
def dog_stage(row):
    """
    Returns the dog stage based on the values of other columns
    """
    stages = ['doggo','floofer','pupper','puppo']
    for stage in stages:
        if row[stage].lower().strip() == stage:
            return stage
    return np.NaN

df_archive_clean['dog_stage'] = df_archive_clean.apply(lambda x: dog_stage(x), axis=1)

#### Test

In [33]:
df_archive_clean['dog_stage'].value_counts(dropna=False)

NaN        1759
pupper      220
doggo        83
puppo        23
floofer       9
Name: dog_stage, dtype: int64

#### Define

* Remove the  columns ('doggo', 'floofer', 'pupper', 'puppo') as they are no longer relevant and now combined into one column `dog_stage`

#### Code

In [34]:
cols = ['doggo', 'floofer', 'pupper', 'puppo']

df_archive_clean.drop(cols, axis=1, inplace=True)

#### Test

In [35]:
df_archive_clean.columns

Index(['tweet_id', 'timestamp', 'source', 'text', 'expanded_urls',
       'rating_numerator', 'rating_denominator', 'name', 'dog_stage'],
      dtype='object')

#### Define

- Fixing the 'rating_numerator' and 'rating_denominator' columns by addressing the possible error causes:
    - for the pack of dogs, the raing denominator is usually a multiple of 10. if that's the case, a factor is calculated to return the rating back to the base of 10.
    - for the errors in extractions of data when the text sontains more than one '/' character. The text is searched using regex for a valid rating with a denominator of 10 and then the rating numerator and denominator are extracted as a result.
    - if there's only one '/' character in the text, a base rating is assumed of 10/10.
    - for the decimal point ratings, the integer value before the decimal point is extracted and considered as the valid rating numerator.
    - only two teets were left with exceptional high scores. They were not changes as they were valid ratings. One for a real dog and another for the famous artist "Snoop Dog". 

#### Code

In [36]:
def fix_rating(row):
    """
    Returns the dataframe row but with fixed rating numerator and denominator
    """
    if row['rating_denominator'] != 10:
        if row['rating_denominator'] % 10 ==0:
            num = row['rating_denominator']/10
            row['rating_numerator']=int(row['rating_numerator']/num)
            row['rating_denominator'] = 10
        else:
            prog = re.compile('../..')
            results = re.findall(prog, row['text'])
            for result in results:
                if result[-2:] == '10':
                    row['rating_numerator']= int(result[:2])
                    row['rating_denominator'] = 10
                    results.remove(result)
            if len(results):
                row['rating_numerator']= 10
                row['rating_denominator'] = 10
                
    else:
        prog = re.compile('\d*\.\d*/10')
        results = re.findall(prog, row['text'])
        for result in results:
            if result[-2:] == '10':
                try:
                    row['rating_numerator'] = int(result.split('.')[0].strip())
                except:
                    row['rating_numerator'] = 10
        
    
    return row

    
df_archive_clean = df_archive_clean.apply(lambda x: fix_rating(x), axis=1)

#### Test

In [37]:
df_archive_clean['rating_denominator'].value_counts()

10    2094
Name: rating_denominator, dtype: int64

In [38]:
df_archive_clean['rating_numerator'].value_counts()

12      487
10      451
11      416
13      287
9       151
8        97
7        51
14       38
5        33
6        32
3        19
4        15
2        10
1         4
420       1
1776      1
0         1
Name: rating_numerator, dtype: int64

#### Define

- The 'name' column is fixed by applying a custom function `fix_name` that tries to find other possibe names in the tweet's text. If no possible name was found in the text the name value is set to null.

#### Code

In [39]:
not_names = ['a', 'the', 'an', 'very', 'one', 'quite', 'just', 'mad', 'actually', 'getting', 'not', 
                 'infuriating', 'space', 'unacceptable', 'his', 'old', 'O', 'life', 'this', 'incredibly', 
                 'such', 'light', 'officially', 'all', 'by', 'my']  

def fix_name(row):
    """
    Returns the row withthe corrected name if found and Null if no name was found
    """
    if row['name'] in not_names or row['name'] == 'None':
        try:
            pattern = re.compile(r'(?:name(?:d)?)\s{1}(?:is\s)?([A-Za-z]+)')
            new_name = re.findall(pattern, row['text'])[0]
            row['name'] = new_name 

        except IndexError:
            row['name'] = np.NaN
    return row


df_archive_clean = df_archive_clean.apply(lambda x: fix_name(x), axis=1)

#### Test

In [40]:
df_archive_clean[['tweet_id','text','name']].query('name in @not_names')

Empty DataFrame
Columns: [tweet_id, text, name]
Index: []

In [41]:
df_archive_clean['name'].value_counts(dropna=False).keys()

Index([      nan, 'Charlie',    'Lucy',  'Cooper',  'Oliver',  'Tucker',
         'Penny',    'Lola',   'Sadie', 'Winston',
       ...
        'Farfle', 'Strider', 'Eleanor',    'Nico',    'Kial',  'Willow',
        'Spanky',    'Kara',  'Vinnie',  'Stuart'],
      dtype='object', length=955)

#### Define

- The tweet's source is extracted from the 'source' column by extracting the inner text of the html 'a' tag using regex. 

#### Code

In [42]:
def fix_source(row):
    """
    Returns the extracted string from the html tag
    """
    pattern = re.compile(r'(?:(?!>).)*?(?=(?:<)|$)')
    source = re.findall(pattern, row['source'])[1]
    if source:
        return source
    else:
        return row['source']
    
    
df_archive_clean['source'] = df_archive_clean.apply(lambda x: fix_source(x), axis=1)

#### Test

In [43]:
df_archive_clean['source'].value_counts()

Twitter for iPhone     1962
Vine - Make a Scene      91
Twitter Web Client       30
TweetDeck                11
Name: source, dtype: int64

In [44]:
df_archive_clean.sample(5)

tweet_id                 timestamp              source  \
957   751538714308972544 2016-07-08 22:09:27+00:00  Twitter for iPhone   
2091  670783437142401025 2015-11-29 01:56:48+00:00  Twitter for iPhone   
1344  704499785726889984 2016-03-01 02:53:32+00:00  Twitter for iPhone   
304   836380477523124226 2017-02-28 01:00:19+00:00  Twitter for iPhone   
2230  668237644992782336 2015-11-22 01:20:44+00:00  Twitter for iPhone   

                                                   text  \
957   This is Max. She has one ear that's always sli...   
2091  Flamboyant pup here. Probably poisonous. Won't...   
1344  When you wake up from a long nap and have no i...   
304   This is Ava. She just blasted off. Streamline ...   
2230  This is Torque. He served his nickel. Better n...   

                                          expanded_urls  rating_numerator  \
957   https://twitter.com/dog_rates/status/751538714...                10   
2091  https://twitter.com/dog_rates/status/670783437...                 1   
1344  https://twitter.com/dog_rates/status/704499785...                12   
304   https://twitter.com/dog_rates/status/836380477...                12   
2230  https://twitter.com/dog_rates/status/668237644...                10   

      rating_denominator    name dog_stage  
957                   10     Max       NaN  
2091                  10     NaN       NaN  
1344                  10     NaN       NaN  
304                   10     Ava    pupper  
2230                  10  Torque       NaN

## 3.2 Tweet Image Predictions

In [45]:
# copy into a new DataFrame to be cleaned
df_pred_clean = df_pred.copy()

#### Define

* Convert the dataframe shape such that each row represents a prediction using the panas `wide_to_long` function.

#### Code

In [46]:
# Renaming the dataset columns
cols = ['tweet_id', 'jpg_url', 'img_num', 
       'prediction_1', 'confidence_1', 'breed_1',
       'prediction_2', 'confidence_2', 'breed_2',
       'prediction_3', 'confidence_3', 'breed_3']

df_pred_clean.columns = cols

df_pred_clean = pd.wide_to_long(df_pred_clean, stubnames=['prediction', 'confidence', 'breed'], 
    i=['tweet_id', 'jpg_url', 'img_num'], j='prediction_level', sep="_").reset_index()

#### Test

In [47]:
df_pred_clean.sample(5)

tweet_id                                          jpg_url  \
3958  756275833623502848  https://pbs.twimg.com/media/Cn7U2xlW8AI9Pqp.jpg   
4963  810254108431155201  https://pbs.twimg.com/media/Cz6Z0DgWIAAfdvp.jpg   
888   671362598324076544  https://pbs.twimg.com/media/CVEouDRXAAEe8mt.jpg   
1882  680836378243002368  https://pbs.twimg.com/media/CXLREjOW8AElfk6.jpg   
482   668902994700836864  https://pbs.twimg.com/media/CUhruUgUAAAa8FQ.jpg   

      img_num  prediction_level                      prediction  confidence  \
3958        1                 2                   Irish_terrier    0.086981   
4963        1                 2  American_Staffordshire_terrier    0.261233   
888         1                 1                             tub    0.393616   
1882        3                 2               Shetland_sheepdog    0.160669   
482         1                 3                Blenheim_spaniel    0.028004   

      breed  
3958   True  
4963   True  
888   False  
1882   True  
482    True

## 3.3 Twitter API

In [48]:
# copy into a new DataFrame to be cleaned
df_api_clean = df_api.copy()

#### Define

* Convert the 'tweet_id', 'retweet_count' and 'favorite_count' columns to integer instead of string.

#### Code

In [49]:
df_api_clean['tweet_id'] = df_api_clean['tweet_id'].astype(int)
df_api_clean['retweet_count'] = df_api_clean['retweet_count'].astype(int)
df_api_clean['favorite_count'] = df_api_clean['favorite_count'].astype(int)

#### Test

In [50]:
df_api_clean.dtypes

tweet_id          int64
retweet_count     int64
favorite_count    int64
dtype: object

## 3.4 Merging Data

#### Define

* Merge the `df_archive_clean` and the `df_api_clean` based on the 'tweet_id' column.

#### Code

In [51]:
df_merged = pd.merge(df_archive_clean, df_api_clean, on='tweet_id')

#### Test

In [52]:
df_merged.sample(3)

tweet_id                 timestamp              source  \
644  766423258543644672 2016-08-18 23:55:18+00:00  Twitter for iPhone   
708  680191257256136705 2015-12-25 01:00:07+00:00  Twitter for iPhone   
868  674790488185167872 2015-12-10 03:19:24+00:00  Twitter for iPhone   

                                                  text  \
644  This is Shadoe. Her tongue flies out of her mo...   
708  Here's a sleepy Christmas pupper 11/10 https:/...   
868    ER... MER... GERD 13/10 https://t.co/L1puJISV1a   

                                         expanded_urls  rating_numerator  \
644  https://twitter.com/dog_rates/status/766423258...                 9   
708  https://twitter.com/dog_rates/status/680191257...                11   
868  https://twitter.com/dog_rates/status/674790488...                13   

     rating_denominator    name dog_stage  retweet_count  favorite_count  
644                  10  Shadoe       NaN           1546            5902  
708                  10     NaN    pupper            479            2080  
868                  10     NaN       NaN            235            1041

## 3.5 Storing Data

export the df_merged file to a new file 'twitter_archive_clean.csv'

In [53]:
df_merged.to_csv('twitter_archive_master.csv',index=False)

export the cleaned image predictions DataFrame to 'image_predictions_clean.csv' 

In [54]:
df_pred_clean.to_csv('image_predictions_clean.csv',index=False)

The exported files are then analyzed and visualized in a separate file [act_report.html](act_report.html)